### PR TITLE
Add EIP: P2P Escrowed Governance Incentives

### PR DESCRIPTION
**Pull request by jhweintraub**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

**Comment by abcoathup** [comment]
EIP editors, is this numbered 6504 due to closed PR #6504?

**Comment by jhweintraub** [comment]
I closed that PR in order to make more local edits and clean up the change history.

**Comment by abcoathup** [comment]
> That was me. I closed that PR in order to make more local edits and clean up the change history. Sorry if that screwed up the process.

Generally EIPs/ERCs are numbered after the first PR #, but will need to wait and see what the EIP editors assign.

**Review by lightclient — COMMENTED** [review]
6506 is fine

File: EIPS/eip-6506.md [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+
+1. Current bribe schemes for votes based on pro-rata distribution are economically innefficient and result in worse outcomes for voters. For systems like Votium or Hidden-Hand, If Alice votes on a proposal with the expectation of receiving $10 in bribes, they can just be backrun by a larger voter, diluting their share of the pool. It may no longer be economical to make the decision they did. Using an OTC mechanisms is more efficient because the amount is “locked in” when the bribe is made and the recipient has much more concrete assurances on which to base their decision. These protocols are also centralized, relying on a central authority to accept and redistribute rewards fairly. Whenever possible, centralization should be avoided.
+
+2. The lack of an existing standard means that parties are relying entirely on trust in one-another to obey. Bob has to trust Alice to pay out and Alice has to trust Bob to vote. Even if the two of them were to use an escrow contract, it may have flaws like relying on a trusted third-party, or simply that it is outside the technical reach of both parties.
+
+
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:26]:
```suggestion
```

File: EIPS/eip-6506.md [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+The key words "BRIBE" and "INCENTIVE" are to be interpreted as the transfer of a digital-asset(s) from user A to user B in exchange for an assurance that user B will vote in a specific-direction, on a specific proposal, for a specified-DAO. If user B does not honor the arrange, the digital-asset(s) will be returned to user A.
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:37]:
```suggestion
The key words "BRIBE" and "INCENTIVE" are to be interpreted as the transfer of a digital-asset(s) from user A to user B in exchange for an assurance that user B will vote in a specific-direction, on a specific proposal, for a specified-DAO. If user B does not honor the arrangement, the digital-asset(s) will be returned to user A. 
```

File: EIPS/eip-6506.md [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+The key words "BRIBE" and "INCENTIVE" are to be interpreted as the transfer of a digital-asset(s) from user A to user B in exchange for an assurance that user B will vote in a specific-direction, on a specific proposal, for a specified-DAO. If user B does not honor the arrange, the digital-asset(s) will be returned to user A. 
+
+The key words "BRIBER", "INCENTIVIZER", and "SENDER" shall refer to the user A offering monetrary compensation to user B. "RECIPIENT", "VOTER", and "INCENTIVIZEE" herein refer to user B whom shall formally receive their compensation upon the termination of the agreement.
+
+The key word "VOTE" shall be interpreted as the casting of a ballot in any kind of governance system which uses accounts as a form of permissions.
+
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:42]: You have no content in your specification section. This is where the meat of your proposal should go. At the very least, you'll want to move your interface from the reference implementation up here (along with any rules that implementations must follow.)

File: EIPS/eip-6506.md [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+2. Lack of transparency - Decisionmaking will occur behind closed doors as the true intentions of voters is unclear, and votes that should pass may fail, or vice-versa. The will of the community will not be honored.
+
+## Backwards Compatibility
+
+No backward compatibility issues found.
+
+
+## Reference Implementation
+
+Contracts wishing to implement such a standard must implement the following interface
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:84]: Yep, it looks like most of this should move into Specification. Use the Reference Implementation for an example contract that implements `GovernanceIncentiveController`.

File: EIPS/eip-6506.md [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+  stateMutability: payable
+
+  inputs:
+    - name: incentiveId
+      type: bytes32
+    - name: incentiveInfo
+      type: bytes calldata
+```
+
+#### claimIncentive
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:185]: You should use backticks for any code in headings, for example:

```markdown
#### `claimIncentive`
```

File: EIPS/eip-6506.md [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+      type: address
+    - name: dismissed
+      indexed: true
+      type: bool
+```
+
+
+## Security Considerations
+
+EIP-TODO is a standard intended to work with existing governance systems. Any potential issue with existing governance may represent a potential attack on this as well. This includes voting-weight manipulation, vote forgery, verification discrepancies etc. All systems in which this EIP is integrated with should be properly audited for maximum security, as any issues may result in improper distribution of these governance incentives.
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:487]:
```suggestion
This proposal is a standard intended to work with existing governance systems. Any potential issue with existing governance may represent a potential attack on this as well. This includes voting-weight manipulation, vote forgery, verification discrepancies etc. All systems in which this EIP is integrated with should be properly audited for maximum security, as any issues may result in improper distribution of these governance incentives.
```

**Comment by jhweintraub** [comment]
I have made the requested revisions alongside various syntatic changes made during development. An official reference implementation was included in the Specifications section, and can be found at the following repo

https://github.com/jhweintraub/ERC6506

File: EIPS/eip-6506.md [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+The key words "BRIBE" and "INCENTIVE" are to be interpreted as the transfer of a digital-asset(s) from user A to user B in exchange for an assurance that user B will vote in a specific-direction, on a specific proposal, for a specified-DAO. If user B does not honor the arrangement, the digital-asset(s) will be returned to user A. 
+
+The key words "BRIBER", "INCENTIVIZER", and "SENDER" shall refer to the user A offering monetrary compensation to user B. "RECIPIENT", "VOTER", and "INCENTIVIZEE" herein refer to user B whom shall formally receive their compensation upon the termination of the agreement.
+
+The key word "VOTE" shall be interpreted as the casting of a ballot in any kind of governance system which uses accounts as a form of permissions.
+
+Contracts wishing to implement such a standard must implement the following interface
+
+```
+
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:45]:
````suggestion
```solidity
````

File: EIPS/eip-6506.md [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+  function modifyClaimer(address claimer, bool designation) external;
+
+  //Dispute Mechanism
+  function beginDispute(bytes32 incentiveId, bytes memory disputeInfo) external payable;
+
+  function resolveDispute(bytes32 incentiveId, bytes memory disputeResolutionInfo) external returns (bool isDismissed);
+
+}
+
+
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:91]:
```suggestion
```

File: EIPS/eip-6506.md [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+#### `incentivize`
+
+The function where an incentivizer should commit to the details of their incentive. The commitment value can be calculated off-chain or calculated on-chain in a full transparency system. The function should take the input data from `incentiveInfo` and create store a new `incentive` object in the mapping incentives. If OUC is enabled, then only incentivizer and timestamp information need be public, everything else should be left as zero.
+
+Function should account for fees taken from user at deposit. If fees are present, then `incentivize` should take them up front. This is to ensure that the amount quoted to a recipient is *at least* as much as they would receive.
+
+MUST emit the `incentiveSent` event
+
+```
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:131]:
````suggestion
```yaml
````

File: EIPS/eip-6506.md [hunk 10]
```diff
@@ -0,0 +1,491 @@
...
+
+ `function verifyVote(bytes32 incentive, bytes memory voteInfo) public view returns (bool isVerifiable);`
+
+ Function used to determine if the voter for `incentive` should receive the incentive originally committed to. 
+
+ Functions may use whatever scheme they like to determine this information. Necesarry data should be encoded and passed through `voteInfo`. 
+
+ MUST return `false` if `voteInfo` indicates that `recipient` did not vote in the direction committed to by `incentive`, and true otherwise.
+
+```
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:213]:
````suggestion
```yaml
````

File: EIPS/eip-6506.md [hunk 11]
```diff
@@ -0,0 +1,491 @@
...
+      type: bool
+    - name: proofData
+      type: bytes memory
+```
+
+#### `modifyClaimer`
+
+Function changing the designation of an address as being approved to claim a bribe on behalf of another user. Only an approved claimer should be able to claim the incentive on behalf of the user which approved them. 
+
+```
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:235]:
````suggestion
```yaml
````

File: EIPS/eip-6506.md [hunk 12]
```diff
@@ -0,0 +1,491 @@
...
+
+A function which is used to resolve pending disputes over `incentiveId`. The exact mechanism shall be specified by the developers.
+
+MUST return false, and be *"dismissed"*, if the mechanisms resolves the dispute in favor of the defendant `(recipient)`, by showing they did honor the incentive of `incentiveId`. If the dispute is *"confirmed"*, then the function should return true. 
+
+MUST transfer funds committed to by `incentivizer` to `recipient` if dispute is `dismissed` and return `funds + fee + bond` to the `plaintiff`. If dismissed, the distribution of the bond shall be at the discretion of the developers. This may including burning, awarding to the defendant, or donating to a community treasury.
+
+MUST emit the event `disputeResolved` on successful resolution.
+
+```
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:290]:
````suggestion
```yaml
````

File: EIPS/eip-6506.md [hunk 13]
```diff
@@ -0,0 +1,491 @@
...
+
+### Events
+
+#### incentiveSent
+
+`incentivizer` has bribed `recipient` `amount` of `token` for some information. 
+
+If system is private then recipient, amount, and `token` may be left as zero.
+
+```
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:314]:
````suggestion
```yaml
````

File: EIPS/eip-6506.md [hunk 14]
```diff
@@ -0,0 +1,491 @@
...
+      type: bytes memory
+  
+  outputs: 
+    - name isDismissed
+      type: bool
+```
+
+### Events
+
+#### incentiveSent
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:308]:
```suggestion
#### `incentiveSent`
```

File: EIPS/eip-6506.md [hunk 15]
```diff
@@ -0,0 +1,491 @@
...
+    - name: amount
+      indexed: true
+      type: uint256
+    - name: recipient
+      indexed: true
+      type: address
+```
+
+
+#### incentiveClaimed
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:334]:
```suggestion
#### `incentiveClaimed`
```

File: EIPS/eip-6506.md [hunk 16]
```diff
@@ -0,0 +1,491 @@
...
+      type: address
+    - name: amount
+      indexed: true
+      type: uint256
+    - name: data
+      indexed: false
+      type: bytes
+```
+
+#### modifiedClaimer
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:357]:
```suggestion
#### `modifiedClaimer`
```

File: EIPS/eip-6506.md [hunk 17]
```diff
@@ -0,0 +1,491 @@
...
+
+  A new `claimer` was either whitelisted by `recipient` or blacklisted.
+
+```
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:361]:
````suggestion
```yaml
````

File: EIPS/eip-6506.md [hunk 18]
```diff
@@ -0,0 +1,491 @@
...
+      type: address
+    - name: claimer
+      indexed: false
+      type: address
+    - name: direction
+      indexed: false
+      type: bool
+```
+
+#### incentiveReclaimed
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:377]:
```suggestion
#### `incentiveReclaimed`
```

File: EIPS/eip-6506.md [hunk 19]
```diff
@@ -0,0 +1,491 @@
...
+
+  An `incentivizer` is reclaiming `incentiveId`, and outing the noncompliance of `voter`
+
+```
```
Comment (SamWilsn) [inline EIPS/eip-6506.md:381]:
````suggestion
```yaml
````

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**